File: android-src/FFmpegPlugin/FFmpegPlugin.java
package com.audioshiftpro.app;

import com.arthenica.ffmpegkit.FFmpegKit;
import com.arthenica.ffmpegkit.FFmpegKitConfig;
import com.arthenica.ffmpegkit.ReturnCode;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "FFmpegConverter")
public class FFmpegPlugin extends Plugin {

    @PluginMethod
    public void convert(PluginCall call) {
        String command = call.getString("command");
        if (command == null || command.isEmpty()) {
            call.reject("Missing command parameter");
            return;
        }

        final String cmd = command;
        new Thread(() -> {
            try {
                FFmpegKitConfig.enableStatisticsCallback(stats -> {
                    JSObject progress = new JSObject();
                    progress.put("time", stats.getTime());
                    progress.put("size", stats.getSize());
                    progress.put("speed", stats.getSpeed());
                    notifyListeners("ffmpegProgress", progress);
                });

                com.arthenica.ffmpegkit.FFmpegSession session = FFmpegKit.execute(cmd);
                JSObject ret = new JSObject();

                if (ReturnCode.isSuccess(session.getReturnCode())) {
                    ret.put("success", true);
                } else if (ReturnCode.isCancel(session.getReturnCode())) {
                    ret.put("success", false);
                    ret.put("error", "Cancelled");
                } else {
                    ret.put("success", false);
                    ret.put("error", session.getFailStackTrace() != null ? session.getFailStackTrace() : "FFmpeg failed rc=" + session.getReturnCode());
                    ret.put("logs", session.getAllLogsAsString());
                }
                call.resolve(ret);
            } catch (Exception e) {
                call.reject("FFmpeg error: " + e.getMessage());
            }
        }).start();
    }

    @PluginMethod
    public void cancel(PluginCall call) {
        FFmpegKit.cancel();
        call.resolve();
    }
}

File: android-src/FFmpegPlugin/FFmpegPlugin.kt
package com.audioshiftpro.app

import android.Manifest
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.FFmpegKitConfig
import com.arthenica.ffmpegkit.ReturnCode
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "FFmpegConverter")
class FFmpegPlugin : Plugin() {

    private var progressCall: PluginCall? = null

    @PluginMethod
    fun convert(call: PluginCall) {
        val command = call.getString("command")
        if (command.isNullOrEmpty()) {
            call.reject("Missing command parameter")
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Enable statistics for progress tracking
                FFmpegKitConfig.enableStatisticsCallback { stats ->
                    val pct = stats.time  // milliseconds processed
                    notifyListeners("ffmpegProgress", JSObject().apply {
                        put("time", pct)
                        put("size", stats.size)
                        put("bitrate", stats.bitrate)
                        put("speed", stats.speed)
                    })
                }

                val session = FFmpegKit.execute(command)
                val ret = JSObject()

                if (ReturnCode.isSuccess(session.returnCode)) {
                    ret.put("success", true)
                    ret.put("returnCode", 0)
                } else if (ReturnCode.isCancel(session.returnCode)) {
                    ret.put("success", false)
                    ret.put("error", "Cancelled")
                } else {
                    ret.put("success", false)
                    ret.put("error", session.failStackTrace ?: "FFmpeg failed with rc=${session.returnCode}")
                    ret.put("logs", session.allLogsAsString)
                }
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("FFmpeg error: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun cancel(call: PluginCall) {
        FFmpegKit.cancel()
        call.resolve()
    }

    @PluginMethod
    fun getVersion(call: PluginCall) {
        val ret = JSObject()
        ret.put("version", com.arthenica.ffmpegkit.BuildConfig.VERSION)
        call.resolve(ret)
    }
}
